File: C2.19.py
#!usr/bin/python
from __future__ import division
from scipy import signal
import math
import numpy as np
import matplotlib.pyplot as plt

'''In this exercise we are going to study the effect or zero padding has on the periodogram. As we already know is that adding zeros
it doesn't give us gain in frequency resolution, it increases the smoothness and we can better characterize the signal.

define the input sequence as y and then as phi we define the periodogram for given frequency
	y(t,N) 10*np.sin(0.2*math.pi*t) + 5np.sin((0.2 + 1/N)*2*math.pi*t) + e(t), e(t) is white noise'''

#define the phi function as given in the exercise, N is fixed, and w = (2*math.pi*k/N)
#in order to calculate the actuall PSD and compare it with the estimations of the periodogram
def phi(k,N=64):
	return (50*math.pi*(signal.unit_impulse((2*math.pi*k)/N - 0.2*2*math.pi) + signal.unit_impulse((2*math.pi*k)/N + 0.2*2*math.pi)) + \
		12.5*math.pi*(signal.unit_impulse((2*math.pi*k)/N - (0.2 + 1/N)*2*math.pi) + \
			signal.unit_impulse((2*math.pi*k)/N + (0.2 + 1/N)*2*math.pi)) + 1)

#define the input signal
def y(t,e,N=64):
	return (10*np.sin(0.2*math.pi*t) + 5*np.sin((0.2 + 1/N)*2*math.pi*t) + e[t])


#first we have to define the Gaussian white noise with variance 1 and 0 mean
def whiteNoise():
	wgn = np.random.normal(0,1,64) # white gaussian noise, with 64 elements from 0 to N-1
	return wgn


def fp(N):
	#now in order to make the frequency array w = 2*pi*k/N for k = 0,....N-1
	f = []
	for  k in range(0,N):
		f.append((2*math.pi*k)/N)
	return f

#make the padding
def zeroPadding(signal,length_of_padding):

	#apply zero padding in the request length
	for x in range(length_of_padding):
		signal.append(0)

	return signal

def main():

	#generate the random white Gaussian noise
	e = whiteNoise()

	#generate the signal for t = 0,.....,N-1
	signal = []
	N = 64 # the initial points in time domain
	for t in range(0,N):
		signal.append(y(t,e))

	

	#calculate the psd plot by applying fft to the sequence of the signal
	psd = np.fft.fft(signal)
	
	#make the frequency array
	fxx = fp(len(signal))

	print len(signal)
	
	#calculate psd
	print 'PSD without padding:', (np.real(sum(psd)/len(psd)))

	signal1 = zeroPadding(signal,N)
	psd1 = np.fft.fft(signal1)

	print len(signal1)
	
	#calculate psd
	print 'PSD with padding N:', (np.real(sum(psd1)/len(psd1)))

	signal2 = zeroPadding(signal,3*N)
	psd2 = np.fft.fft(signal2)

	print len(signal2)

	#calculate psd
	print 'PSD with padding 3N:', (np.real(sum(psd2)/len(psd2)))

	signal3 = zeroPadding(signal,5*N)
	psd3 = np.fft.fft(signal3)

	print len(signal3)

	#calculate psd
	print 'PSD with padding 5N:', (np.real(sum(psd3)/len(psd3)))

	signal4 = zeroPadding(signal,7*N)
	psd4 = np.fft.fft(signal4)

	print len(signal4)

	#calculate psd
	print 'PSD with padding 7N:', (np.real(sum(psd4)/len(psd4)))
	

	#plots
	fig1 = plt.figure(1,figsize=(12,6))
	plt.plot(e,color='blue')
	plt.xlabel('Time (s)')
	plt.ylabel('Amplitude')
	plt.title('white Gaussian noise e(t)')
	fig1.savefig('white_noise.png')

	fig2 = plt.figure(2,figsize=(12,6))
	plt.plot(signal,color='blue')
	plt.xlabel('Time (s)')
	plt.ylabel('y(t)')
	plt.title('Signal with padding 7N')
	#fig2.savefig('signal_padding_7N.png')

	fig3 = plt.figure(3,figsize=(12,6))
	plt.plot(fxx,psd,color='blue')
	plt.xlabel('Frequency (Hz)')
	plt.ylabel('PSD')
	plt.title('Periodogram of y(t) without padding')
	#fig3.savefig('periodoram_without_padding.png')

	#because psd is symmetric, plot only the 0-pi part because pi to 2*pi is the same
	fig4 = plt.figure(4,figsize=(12,6))
	plt.plot(fxx[:int(len(signal)/2)],psd[:int(len(signal)/2)],color='blue')
	plt.xlabel('Frequency (Hz)')
	plt.ylabel('PSD')
	plt.title('Periodogram of y(t) with padding 7N')
	#fig4.savefig('periodoram_wit_padding_7N_half.png')
	

	plt.show()

main()
	

File: C2.19.m
whitegaussiannoise = wgn(64,1,0);%generates an array 64x1 elements with 0 power (white gaussian noise with variance(power) 1 watt so 0 dB)
fprintf('Variance of the signal: %f\n', var(whitegaussiannoise))
%Sigma (standard deviation) uses the same formula as when calculating the RMS value 
%of a voltage or current with no dc content. Inspect the two formula and see. 
%Sigma squared is therefore a measure of power.
time = 1:64;%make times axis for 64 samples, one sample per second


%generate our custom signal
N = 64;% where t = 0....N-1 the samples from the time domain
t = (1:N);
y = 10*sin(0.2*pi*t) + 5*sin((0.2 + 1/N)*2*pi*t);

%calculate the periodogram of the signal by applying fourier transform
periodogram = real(fft(y));

%determine the frequency array
frequencyArray = 2*pi*t/N;


%DETERMINE ZERO PADDINGGG!!!!!!!!

%zero padding N
zero_paddedSignalN = [y zeros(1,N)];

%extend the time
zero_paddingTimeN = (1:2*N);

%calculate padded periodogram
zero_paddedPeriodogramN = real(fft(zero_paddedSignalN));

%determine frequencies
zero_paddingFrequencyArrayN = 2*pi*zero_paddingTimeN/(2*N);




%zero padding 3N
zero_paddedSignal3N = [y zeros(1,3*N)];

%extend the time
zero_paddingTime3N = (1:4*N);

%calculate padded periodogram
zero_paddedPeriodogram3N = real(fft(zero_paddedSignal3N));

%determine frequencies
zero_paddingFrequencyArray3N = 2*pi*zero_paddingTime3N/(4*N);




%zero padding 5N
zero_paddedSignal5N = [y zeros(1,5*N)];

%extend the time
zero_paddingTime5N = (1:6*N);

%calculate padded periodogram
zero_paddedPeriodogram5N = real(fft(zero_paddedSignal5N));

%determine frequencies
zero_paddingFrequencyArray5N = 2*pi*zero_paddingTime5N/(6*N);




%zero padding 7N
zero_paddedSignal7N = [y zeros(1,7*N)];

%extend the time
zero_paddingTime7N = (1:8*N);

%calculate padded periodogram
zero_paddedPeriodogram7N = real(fft(zero_paddedSignal7N));

%determine frequencies
zero_paddingFrequencyArray7N = 2*pi*zero_paddingTime7N/(8*N);



%plots
figure(1)
subplot(4,1,1)
plot(time,whitegaussiannoise)
xlabel('Time (sec)')
ylabel('Amplitude') 
title('White Gaussian noise')
subplot(4,1,2)
plot(time,y)
xlabel('Time (sec)')
ylabel('Amplitude') 
title('y(t)')
subplot(4,1,3)
plot(frequencyArray,periodogram);
ylim([-200,400]);  % static limits
xlabel('Frequency (Hz)')
ylabel('PSD [V**2/Hz]') % The power spectral density (PSD) of the signal describes the power present in 
%the signal as a function of frequency, per unit frequency
title('Periodogram of y(t)')
subplot(4,1,4)
plot(frequencyArray(1:32),periodogram(1:32))
ylim([-200,400]);  % static limits
xlabel('Frequency (Hz)')
ylabel('PSD [V**2/Hz]') 
title('Periodogram of y(t) non-symmetric')

%figure 2 in order to plot the padded signals
figure(2)
subplot(4,1,1)
plot(time,whitegaussiannoise)
xlabel('Time (sec)')
ylabel('Amplitude') 
title('White Gaussian noise')
subplot(4,1,2)
plot(zero_paddingTimeN,zero_paddedSignalN)
xlabel('Time (sec)')
ylabel('Amplitude') 
title('zero padded N y(t)')
subplot(4,1,3)
plot(zero_paddingFrequencyArrayN,zero_paddedPeriodogramN);
xlabel('Frequency (Hz)')
ylabel('PSD [V**2/Hz]') 
title('Periodogram of zero padded N y(t)')
subplot(4,1,4)
plot(zero_paddingFrequencyArrayN(1:N),zero_paddedPeriodogramN(1:N))
xlabel('Frequency (Hz)')
ylabel('PSD [V**2/Hz]') 
title('Periodogram of zero padded N y(t) non-symmetric')


%figure 2 in order to plot the padded signals
figure(3)
subplot(4,1,1)
plot(time,whitegaussiannoise)
xlabel('Time (sec)')
ylabel('Amplitude') 
title('White Gaussian noise')
subplot(4,1,2)
plot(zero_paddingTime3N,zero_paddedSignal3N)
xlabel('Time (sec)')
ylabel('Amplitude') 
title('zero padded 3N y(t)')
subplot(4,1,3)
plot(zero_paddingFrequencyArray3N,zero_paddedPeriodogram3N);
xlabel('Frequency (Hz)')
ylabel('PSD [V**2/Hz]') 
title('Periodogram of zero padded 3N y(t)')
subplot(4,1,4)
plot(zero_paddingFrequencyArray3N(1:(2*N)),zero_paddedPeriodogram3N(1:(2*N)))
xlabel('Frequency (Hz)')
ylabel('PSD [V**2/Hz]') 
title('Periodogram of zero padded 3N y(t) non-symmetric')


%figure 2 in order to plot the padded signals
figure(4)
subplot(4,1,1)
plot(time,whitegaussiannoise)
xlabel('Time (sec)')
ylabel('Amplitude') 
title('White Gaussian noise')
subplot(4,1,2)
plot(zero_paddingTime5N,zero_paddedSignal5N)
xlabel('Time (sec)')
ylabel('Amplitude') 
title('zero padded 5N y(t)')
subplot(4,1,3)
plot(zero_paddingFrequencyArray5N,zero_paddedPeriodogram5N);
xlabel('Frequency (Hz)')
ylabel('PSD [V**2/Hz]') 
title('Periodogram of zero padded 5N y(t)')
subplot(4,1,4)
plot(zero_paddingFrequencyArray5N(1:(3*N)),zero_paddedPeriodogram5N(1:(3*N)))
xlabel('Frequency (Hz)')
ylabel('PSD [V**2/Hz]') 
title('Periodogram of zero padded 5N y(t) non-symmetric')


%figure 2 in order to plot the padded signals
figure(5)
subplot(4,1,1)
plot(time,whitegaussiannoise)
xlabel('Time (sec)')
ylabel('Amplitude') 
title('White Gaussian noise')
subplot(4,1,2)
plot(zero_paddingTime7N,zero_paddedSignal7N)
xlabel('Time (sec)')
ylabel('Amplitude') 
title('zero padded 7N y(t)')
subplot(4,1,3)
plot(zero_paddingFrequencyArray7N,zero_paddedPeriodogram7N);
xlabel('Frequency (Hz)')
ylabel('PSD [V**2/Hz]') 
title('Periodogram of zero padded 7N y(t)')
subplot(4,1,4)
plot(zero_paddingFrequencyArray7N(1:(4*N)),zero_paddedPeriodogram7N(1:(4*N)))
xlabel('Frequency (Hz)')
ylabel('PSD [V**2/Hz]') 
title('Periodogram of zero padded 7N y(t) non-symmetric')
               
               
                            
                            